scripts/linking.py

```diff
@@ -385,72 +384,46 @@ def eval_candidate_generation(examples: List[data_util.MedMentionExample],
-
-                    doc_candidates = set()
-                    doc_golds = set(gold_umls_ids)
-
-                    # it is possible that a spacy model does not find any entities in an abstract
-                    if ner_mentions == []:
-                        doc_entity_missed_count += len(doc_golds)
-                        all_golds_per_doc_set += list(doc_golds)
-                        all_golds += gold_umls_ids
-                        continue
+                    mention_texts = [ent.text for ent in doc.ents]
 
-                    batch_candidate_neighbor_ids = candidate_generator.generate_candidates(ner_mentions, k)
-                    for ner_mention, candidate_neighbor_ids in zip(ner_mentions, batch_candidate_neighbor_ids):
-                        # Keep only canonical entities for which at least one mention has a score less than the threshold.
-                        filtered_ids = {k: v for k, v in candidate_neighbor_ids.items() if any([z[1] <= threshold for z in v])}
-                        num_candidates.append(len(candidate_neighbor_ids))
-                        num_filtered_candidates.append(len(filtered_ids))
-
-                        doc_candidates.update(filtered_ids)
+                batch_candidate_neighbor_ids = candidate_generator.generate_candidates(mention_texts, k)
 
-                    for gold_entity in entities:
+                for candidate_neighbor_ids in batch_candidate_neighbor_ids:
+                    # Keep only canonical entities for which at least one mention has a score less than the threshold.
+                    filtered_ids = {k: v for k, v in candidate_neighbor_ids.items() if any([z[1] <= threshold for z in v])}
+                    num_candidates.append(len(candidate_neighbor_ids))
+                    num_filtered_candidates.append(len(filtered_ids))
+                    doc_candidates.update(filtered_ids)
+
+                for i, gold_entity in enumerate(entities):
+                    if use_gold_mentions:
+                        candidates = batch_candidate_neighbor_ids[i]  # for gold mentions, len(entities) == len(batch_candidate_neighbor_ids)
+                    else:
+                        # for each gold entity, search for a corresponding predicted entity that has the same span
                         span_from_doc = doc.char_span(gold_entity.start, gold_entity.end)
                         candidates = {}
                         for j, predicted_entity in enumerate(ner_entities):
                             if predicted_entity == span_from_doc:
                                 candidates = batch_candidate_neighbor_ids[j]
                                 break
 
-                        # Keep only canonical entities for which at least one mention has a score less than the threshold.
-                        filtered_ids = {k: v for k, v in candidates.items() if any([z[1] <= threshold for z in v])}
-                        num_candidates.append(len(candidates))
```

Comment:
@danielkingai2, I deleted this line and the following two because these counts have already been updated in the loop above. I wanted to double check with you that this is correct

Reply:
Ah, yes. that is correct.